File: groove/groove-node/tape.ts
import { gunzipSync, gzipSync } from "zlib";


interface TapeRequest {
    url: string
    method: string
    headers: Record<string, string[]>
    body: Buffer
}

interface TapeResponse {
    status: number
    headers: Record<string, string[]>
    body: Buffer
}

interface TapeRecord {
    request: TapeRequest
    response: TapeResponse
}

export class TapeSession {
    records: TapeRecord[];

    constructor() {
        this.records = [];
    }

    async readFromServer(contents: Buffer) {
        // un-gzip and un-json the blob
        const uncompressed = gunzipSync(contents);
        const jsonPayload = JSON.parse(uncompressed.toString());

        // Un-base64 the bodies
        this.records = jsonPayload.map((record: any) => {
            return {
                ...record,
                request: {
                    ...record.request,
                    body: Buffer.from(record.request.body, 'base64'),
                },
                response: {
                    ...record.response,
                    body: Buffer.from(record.response.body, 'base64'),
                }
            }
        });
    }

    toServer() : Buffer {
        // Base64 the bodies
        const jsonPayload = this.records.map((record) => {
            return {
                ...record,
                request: {
                    ...record.request,
                    body: record.request.body.toString('base64'),
                },
                response: {
                    ...record.response,
                    body: record.response.body.toString('base64'),
                }
            }
        });

        // Json and gzip the blob
        const compressed = gzipSync(JSON.stringify(jsonPayload));
        return compressed;
    }
}

File: groove/groove-python/groove/tape.py
from base64 import b64decode, b64encode
from gzip import compress, decompress
from json import dumps, loads

from pydantic import validator

from groove.models import GrooveModelBase


class TapeRequest(GrooveModelBase):
    url: str
    method: str
    headers: dict[str, list[str]]
    body: bytes

    @validator("body")
    def validate_body(cls, value):
        return b64decode(value)


class TapeResponse(GrooveModelBase):
    status: int
    headers: dict[str, list[str]]
    body: bytes

    @validator("body")
    def validate_body(cls, value):
        return b64decode(value)


class TapeRecord(GrooveModelBase):
    request: TapeRequest
    response: TapeResponse

    class Config:
        json_encoders = {
            # Assume that body bytes should always be encoded as base64 strings
            bytes: lambda value: b64encode(value).decode(),
        }


class TapeSession(GrooveModelBase):
    records: list[TapeRecord]

    @classmethod
    def from_server(cls, data: bytes):
        raw_records = loads(decompress(data))
        return cls(records=raw_records)

    def to_server(self) -> bytes:
        return compress(
            dumps([
                # json_encoders doesn't operate on list items, must iterate manually
                # https://github.com/pydantic/pydantic/issues/4085
                loads(record.json())
                for record in self.records
            ]).encode()
        )